Set job back to 'pending' in case of exception

mdietrichc2c opened the pull request:
This PR is for issue #148 in 7.0 ; the goal is to set back the job in state `pending` if execution of the asynchronous job raised an exception (timeout or other).

mdietrichc2c committed:
Set job back to 'pending' in case of exception

lmignon commented on connector/jobrunner/runner.py:
```diff
@@ -98,19 +98,33 @@
+        with closing(conn.cursor()) as cr:
+            cr.execute("UPDATE queue_job SET state=%s, "
+                       "date_enqueued=NULL, date_started=NULL, worker_uuid=NULL "
+                       "WHERE uuid=%s", (PENDING, job_uuid))
```
we should add 'state='enqueued'` to the where clause

guewen replied:
Yes, because if the controller changes the state to `started` and we get a timeout or no response, the jobrunner could reset a running job to `enqueued` (not sure that could really happen but anyway that's safer).

guewen commented on connector/jobrunner/runner.py:
```diff
@@ -98,19 +98,33 @@
             requests.get(url, timeout=1)
         except requests.Timeout:
-            pass
+            _logger.exception("timeout in GET %s", url)
```
Could you change this log level to `error`? We don't need the traceback on this one.

mdietrichc2c commented:
@lmignon @guewen thanks for the feedback ; I did the requested changes.

guewen commented:
Thanks for the fix! :+1:

lmignon commented:
@mdietrichc2c Thank you for the changes! The code looks good to me. :+1:

laetitia-gangloff commented:
It seems ok :+1: 
Thank you !

guewen commented:
Not yet ready to merge, still we have an issue with this fix.

lmignon commented:
@guewen, on our side @laetitia-gangloff is also testing these changes and that's why I don't have merged these changes despite the 3 thumbs up.

laetitia-gangloff commented:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(_self.__args, *_self.__kwargs)
  File "/home/lga/project/openerp/addons-connector/connector/jobrunner/runner.py", line 129, in urlopen
    set_job_pending()
  File "/home/lga/project/openerp/addons-connector/connector/jobrunner/runner.py", line 111, in set_job_pending
    "WHERE uuid=%s and state=%s", (PENDING, job_uuid, ENQUEUED)
ProgrammingError: column "worker_uuid" of relation "queue_job" does not exist
LINE 1: ...'pending', date_enqueued=NULL, date_started=NULL, worker_uui...

guewen commented:
:-1: until it is fixed

laetitia-gangloff commented:
I think ", worker_uuid=NULL" can be removed...
I did the following test :

start odoo
set a lot of job in the queue (7000)
start runner
wait 30 seconds
stop odoo
remove started job from my database
stop runner
start odoo
start runner

And all seems ok.

mdietrichc2c commented:
I have another issue: even when removeing the `autocommit` and adding `cr.commit()` at the end, the commit does not happen on an enqueued job (I created a timeout for testing).

sbidoul commented on connector/jobrunner/runner.py:
```diff
@@ -98,19 +98,35 @@
         except requests.Timeout:
-            pass
+            _logger.error("timeout in GET %s", url)
+            set_job_pending()
```
Timeout is a normal behaviour for long running jobs. So I would not log this as an error.

mdietrichc2c commented:
Turns out the error I had was due to my test (instant timeout => job was enqueued, set as pending, then running, enqueued, timeout...)

@sbidoul @laetitia-gangloff I corrected both points : removed the absent column from the SQL request , and the message is now logged as a `debug` message.

mdietrichc2c commented:
Actually, after discussing it with @guewen , every job longer than a second will log a timeout message ; it was decided to remove them for timeouts, and I rebased the whole thing in one commit.

guewen commented:
Thanks a lot @mdietrichc2c :-) 
I lift my thumb down :+1:

lmignon commented on connector/jobrunner/runner.py:
```diff
@@ -98,19 +98,34 @@
             requests.get(url, timeout=1)
         except requests.Timeout:
-            pass
+            set_job_pending()
```
What will happen if a job is longer than 1 min? The state will be set to 'pending' and the running job will fails with a `concurrent update error`?

mdietrichc2c replied:
Wouldn't the started job have `started` as a state, and hence not be reset as `pending` ?

guewen replied:
Why? The state of the job is set to `started` when `/runjob` is called. From my understanding, this use case only cover a timeout that happen before we could hit the controller. If we hit the controller, the job is set to `started` so the job won't be reset to `pending`.

lmignon replied:
@mdietrichc2c you're right

lmignon commented:
Thank you @mdietrichc2c . I can now comment my :+1:  with 'Code review and functional test'

laetitia-gangloff commented:
I redo the same test than before, and it works as expected :+1:
